components/lib/state/src/main/java/mozilla/components/lib/state/Store.kt

```diff
@@ -0,0 +1,108 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.lib.state
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.launch
+import java.util.concurrent.Executors
+
+typealias Reducer<S, A> = (S, A) -> S
+typealias Observer<S> = (S) -> Unit
+
+/**
+ * A generic store holding an immutable [State].
+ *
+ * The [State] can only be modified by dispatching [Action]s which will create a new state and notify all registered
+ * [Observer]s.
+ *
+ * @param initialState The initial state until a dispatched [Action] creates a new state.
+ * @param reducer A function that gets the current [State] and [Action] passed in and will return a new [State].
+ */
+open class Store<S : State, A: Action>(
+    initialState: S,
+    private val reducer: Reducer<S, A>
+) {
+    private val storeContext = Executors.newSingleThreadExecutor().asCoroutineDispatcher()
+    private val storeScope = CoroutineScope(storeContext)
+    private val subscriptions: MutableList<Subscription<S, A>> = mutableListOf()
```

Comment:
Use a `MutableSet` instead since the subscriptions are removed by value, not index